stdlib/public/core/CollectionAlgorithms.swift.gyb

```diff
@@ -142,149 +142,84 @@ orderingExplanation = """\
-  ///   argument should be ordered before its second argument; otherwise,
-  ///   `false`.
-  /// - Returns: A pivot index, such that every element before the pivot is
-  ///   ordered before every element at or above the pivot, using
-  ///   `isOrderedBefore` to determine the relative order of any two elements.
-  ///   The returned pivot is equal to the collection's end index only if the
-  ///   collection is empty.
-  ///
-  /// - SeeAlso: `partition()`
+extension MutableCollection {
   public mutating func partition(
-    isOrderedBefore: @noescape (${IElement}, ${IElement}) -> Bool
-  ) -> Index
+      by belongsInSecondPartition: @noescape (${IElement}) throws -> Bool
+  ) rethrows -> Index {
 
-%   else:
+    var pivot = startIndex
+    while true {
+      if pivot == endIndex {
+        return pivot
+      }
+      if try belongsInSecondPartition(self[pivot]) {
+        break
+      }
+      formIndex(after: &pivot)
+    }
 
-  /// Reorders the elements in the collection and returns a pivot index.
-  ///
-  /// This method is typically one step of a sorting algorithm. A collection is
-  /// partitioned around a pivot index when each of the elements before the
-  /// pivot are less than each of the elements at or after the pivot. The
-  /// `partition()` method reorders the elements of the collection and returns
-  /// a pivot index that satisfies this condition.
-  ///
-  /// For example:
-  ///
-  ///     var numbers = [50, 30, 60, 50, 80, 10, 40, 30]
-  ///     let pivot = numbers.partition()
-  ///
-  ///     print(pivot)
-  ///     // Prints "4"
-  ///     print(numbers)
-  ///     // Prints "[10, 30, 30, 40, 50, 80, 50, 60]"
-  ///
-  /// The return value of the call to `numbers.partition()` is the pivot for
-  /// the rearranged `numbers` array. `pivot` divides the collection into two
-  /// subranges, `numbers[0..<pivot]` and `numbers[pivot..<8]`.
-  ///
-  ///     print(numbers[0..<pivot])
-  ///     // Prints "[10, 30, 30, 40]"
-  ///     print(numbers[pivot..<8])
-  ///     // Prints "[50, 80, 50, 60]"
-  ///
-  /// The elements of `numbers` are rearranged so that every element in the
-  /// subrange before `pivot` is less than every element in the subrange
-  /// after.
-  ///
-  /// - Returns: A pivot index, such that every element before the pivot is
-  ///   less than every element at or above the pivot. The returned pivot is
-  ///   equal to the collection's end index only if the collection is empty.
-  ///
-  /// - SeeAlso: `partition(isOrderedBefore:)`
-  public mutating func partition() -> Index
+    var i = index(after: pivot)
+    while i < endIndex {
+      if try !belongsInSecondPartition(self[i]) {
+        swap(&self[i], &self[pivot])
+        formIndex(after: &pivot)
+      }
+      formIndex(after: &i)
+    }
+    return pivot
+  }
+}
 
-%   end
-  {
-    let maybeOffset = _withUnsafeMutableBufferPointerIfSupported {
+extension MutableCollection where Self: BidirectionalCollection {
```

Comment:
Please add a space before the colon.

I'd also be surprised if the existing test compiled, since it probably requires `RandomAccessCollection`.